File: src/models/nsa/slc_attn.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file selected_attention.cpp
 * \brief
 */

#include "interface/operation/operation_impl.h"
#include "interface/operation/operation.h"
#include "interface/function/function.h"
#include "tilefwk/tensor.h"
#include "interface/tensor/logical_tensor.h"
#include "interface/tensor/raw_tensor.h"
#include "tilefwk/tilefwk.h"
#include "tilefwk.h"
#include "interface/program/program.h"
#include "interface/tensor/tensormap.h"
#include "interface/configs/config_manager.h"
#include "interface/configs/config.h"
#include "interface/utils/common.h"
#include "interface/utils/id_gen.h"
#include "interface/utils/log.h"
#include "selected_attention.h"

using namespace npu::tile_fwk;

namespace npu::tile_fwk {
/**
 * normal attention: q=qNope+qRope, kv是连续的
 * input:
    * qNope: [b*s1*n2*g, k_dim] fp16/bf16
    * qRope: [b*s1*n2*g, rope_dim] fp16/bf16
    * kSlc: [b*s1*n2*s2, k_dim+rope_dim], nope与rope在gen_kv_slc中已经合并起来了 fp16/bf16
    * vSlc: [b*s1*n2*s2, v_dim] fp16/bf16
    * kvSlcActSeqs: [b] int32
 * output:
    * attentionOut: [b*s1*n2*g, v_dim] fp32
*/
void SlcAttn(Tensor &qNope, Tensor &qRope, Tensor &kSlc, Tensor &vSlc, Tensor &kvSlcActSeqs, int nQ, int nKv,
    float softmaxScale, Tensor &attentionOut, SaTileShapeConfig tileConfig) {
    auto dtype = qNope->Datatype();
    int dN = qNope->shape[1];
    int dR = qRope->shape[1];
    int group = nQ / nKv;

    int gTile = tileConfig.gTile;
    int s2Tile = tileConfig.sKvTile;
    auto c1Tile = tileConfig.c1TileShape;
    auto v1Tile = tileConfig.v1TileShape;
    auto c2Tile = tileConfig.c2TileShape;
    auto v2Tile = tileConfig.v2TileShape;

    FUNCTION("SA_MAIN", FunctionType::DYNAMIC, {qNope, qRope, kSlc, vSlc, kvSlcActSeqs}, {attentionOut}) {
        /******** tune params ********/
        // Program::GetInstance().GetConfig().SetCubeNBufferMap({});
        // Program::GetInstance().GetConfig().SetL1Reuse(0);
        // Program::GetInstance().GetConfig().SetCopyInThreshold(1 * NUM_1024 * NUM_1024);
        // Program::GetInstance().GetConfig().SetCycleUpperBound(NUM_100000);
        // Program::GetInstance().GetConfig().SetParallelThreshold(NUM_2);
        // Program::GetInstance().GetConfig().SetCubeNBuffer(NUM_2);
        // config::SetOperationConfig("FORCE_COMBINE_AXIS", true);

        /******** attention ********/
        SymbolicScalar batchSizeSym = GetInputShapeDim(kvSlcActSeqs, 0); // b
        SymbolicScalar s1N2GSym = GetInputShapeDim(qNope, 0) / batchSizeSym; // s1n2
        SymbolicScalar s1Sym = s1N2GSym / nQ; // s1
        SymbolicScalar gLoopSym = group / gTile;

        SymbolicScalar s1N2S2Sym = GetInputShapeDim(kSlc, 0) / batchSizeSym; // s1n2s2
        SymbolicScalar n2S2Sym = s1N2S2Sym / s1Sym; // n2s2
        SymbolicScalar n2Sym = nKv;

        LOOP("SA_LOOP_L0_b", FunctionType::DYNAMIC_LOOP, bIdx, LoopRange(0, batchSizeSym, 1)) {
            SymbolicScalar curKvSlcSeq = GetInputDataInt32Dim1(kvSlcActSeqs, bIdx);

            LOOP("SA_LOOP_L1_s1", FunctionType::DYNAMIC_LOOP, s1Idx, LoopRange(0, s1Sym, 1)) { // 每一个S1的k_slc/v_slc都不同
                SymbolicScalar curSeq = std::max(curKvSlcSeq - s1Sym + 1 + s1Idx, 0); // for MTP s1!= 1 casual计算
                curSeq.AsIntermediateVariable();
                SymbolicScalar bnPerBatch = (curSeq + s2Tile - 1) / s2Tile;

                LOOP("SA_LOOP_L2_n2", FunctionType::DYNAMIC_LOOP, n2Idx, LoopRange(0, n2Sym, 1)) { // GQA场景
                    LOOP("SA_LOOP_L3_g", FunctionType::DYNAMIC_LOOP, gIdx, LoopRange(0, gLoopSym, 1)) {
                        int curGTile = gTile;
                        Tensor oiUpdate(DT_FP32, {curGTile, dN}, "oiUpdate");
                        Tensor liUpdate(DT_FP32, {curGTile, 1}, "liUpdate");
                        Tensor miUpdate(DT_FP32, {curGTile, 1}, "miUpdate");

                        SymbolicScalar curOffset = bIdx * s1N2GSym + s1Idx * nQ + n2Idx * group + gIdx * curGTile;
                        std::vector<SymbolicScalar> oiOffset = {curOffset, 0}; // 按最终结果(B*S1*N1,D)进行assemble

                        LOOP("SA_LOOP_L4_s2", FunctionType::DYNAMIC_LOOP, s2Idx, LoopRange(0, bnPerBatch, 1), PowersOf2(1)) {
                            int curS2Tile = s2Tile;
                            SymbolicScalar curKvOffset = bIdx * s1N2S2Sym + s1Idx * n2S2Sym + s2Idx * curS2Tile;

                            ConfigManager::Instance().SetSemanticLabel("Sa");
                            auto qn = DView(qNope, {curGTile, dN}, {curOffset, 0});
                            auto qr = DView(qRope, {curGTile, dR}, {curOffset, 0});
                            Tensor qi(dtype, {curGTile, dN + dR}, "qi");
                            DAssemble(qn, {0, 0}, qi);
                            DAssemble(qr, {0, dN}, qi);

                            auto kj = DViewPad(kSlc, {curS2Tile, dN + dR}, {std::min(curSeq - s2Idx * curS2Tile, curS2Tile), dN + dR},
                                            {curKvOffset, 0}); // kSlc已经合并了rope和nope
                            auto vj = DViewPad(vSlc, {curS2Tile, dN}, {std::min(curSeq - s2Idx * curS2Tile, curS2Tile), dN},
                                            {curKvOffset, 0});

                            // C1
                            Program::GetInstance().GetTileShape().SetCubeTileShapes(
                                {c1Tile[0], c1Tile[1]}, {c1Tile[2], c1Tile[3]}, {c1Tile[4], c1Tile[5]}, true);
                            ConfigManager::Instance().SetSemanticLabel("Sa_QkMM");
                            Program::GetInstance().GetMatrixSize().SetMatrixSize({qi.GetShape()[0], 0, kj.GetShape()[0]});
                            auto sij = Matrix::Matmul<false, true>(DataType::DT_FP32, qi, kj); // (curNTile, dN+dR), (curS2Tile, dN+dR) -> (curNTile, curS2Tile)

                            // V1
                            ConfigManager::Instance().SetSemanticLabel("Sa_Qkvec1");
                            Program::GetInstance().GetTileShape().SetVecTileShapes(v1Tile[0], v1Tile[1]);
                            auto sijScale = MulS(sij, Element(sij->Datatype(), softmaxScale)); // (curNTile, curS2Tile)
                            auto tildaMij = RowMaxSingle(sijScale); // (curNTile, curS2Tile) -> (curNTile, 1)
                            auto tsub = Sub(sijScale, tildaMij);
                            auto tildaPij = Exp(tsub);
                            auto tildaPijF16 = Cast(tildaPij, dtype);
                            auto tildaLij = RowSumSingle(tildaPij); // (nTileCur, s2TileCur) -> (nTileCur, 1)

                            IF (IsLoopBegin(s2Idx, 0)) {
                                // C2
                                Program::GetInstance().GetTileShape().SetCubeTileShapes(
                                    {c2Tile[0], c2Tile[1]}, {c2Tile[2], c2Tile[3]}, {c2Tile[4], c2Tile[5]}, true);
                                ConfigManager::Instance().SetSemanticLabel("Sa_KvMm");
                                Program::GetInstance().GetMatrixSize().SetMatrixSize(
                                    {tildaPijF16.GetShape()[0], tildaPijF16.GetShape()[1], vj.GetShape()[1]});
                                auto oiTmp = Matrix::Matmul<false, false>(DataType::DT_FP32, tildaPijF16, vj);
                                Program::GetInstance().GetTileShape().SetVecTileShapes(v2Tile[0], v2Tile[1]);
                                IF (IsLoopEnd(s2Idx, bnPerBatch)) {
                                    // V2
                                    ConfigManager::Instance().SetSemanticLabel("Sa_KvVec2");
                                    oiUpdate = Div(oiTmp, tildaLij); // (nTileCur, dN) / (nTileCur, 1) -> (nTileCur, dN)
                                    DAssemble(oiUpdate, oiOffset, attentionOut);
                                } ELSE {
                                    oiUpdate = oiTmp;
                                }
                                liUpdate = tildaLij;
                                miUpdate = tildaMij;
                            } ELSE {
                                ConfigManager::Instance().SetSemanticLabel("Sa_UpdateVec2");
                                auto oi = oiUpdate;
                                auto li = liUpdate;
                                auto mi = miUpdate;

                                auto miNew = Maximum(mi, tildaMij); // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)
                                auto t1 = Sub(mi, miNew);           // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)
                                auto t2 = Exp(t1);
                                auto t3 = Sub(tildaMij, miNew); // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)
                                auto t4 = Exp(t3);
                                auto t5 = Mul(t4, tildaLij); // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)
                                auto t6 = Mul(t2, li);       // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)
                                auto liNew = Add(t6, t5);    // (curNTile, 1), (curNTile, 1) -> (curNTile, 1)

                                auto q3 = Mul(oi, t2); // (curNTile, dN), (curNTile, 1) -> (curNTile, dN)
                                Program::GetInstance().GetTileShape().SetCubeTileShapes(
                                    {c2Tile[0], c2Tile[1]}, {c2Tile[2], c2Tile[3]}, {c2Tile[4], c2Tile[5]}, true);
                                ConfigManager::Instance().SetSemanticLabel("Sa_UpdateMM2");
                                Program::GetInstance().GetMatrixSize().SetMatrixSize(
                                    {tildaPijF16.GetShape()[0], tildaPijF16.GetShape()[1], vj.GetShape()[1]});
                                auto q1 = Matrix::Matmul<false, false>(DataType::DT_FP32, tildaPijF16, vj);
                                Program::GetInstance().GetTileShape().SetVecTileShapes(v2Tile[0], v2Tile[1]);
                                auto q2 = Mul(q1, t4);    // (nTileCur, dN), (nTileCur, 1) -> (nTileCur, dN)
                                auto oiTmp = Add(q3, q2); // (nTileCur, dN), (nTileCur, dN) -> (nTileCur, dN)
                                IF (IsLoopEnd(s2Idx, bnPerBatch)) {
                                    oiUpdate = Div(oiTmp, liNew); // (nTileCur, dN) / (nTileCur, 1) -> (nTileCur, dN)
                                    DAssemble(oiUpdate, oiOffset, attentionOut);
                                } ELSE {
                                    oiUpdate = oiTmp;
                                }
                                liUpdate = liNew;
                                miUpdate = miNew;
                            }
                        }
                    }
                }
            }
        }
    }
}
} // namespace npu::tile_fwk

File: python/src/tensor_op/example/slc_attn.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
"""
"""
from dataclasses import dataclass
from typing import List
import logging
import math
import pto

KEY_SUPPORT_DYNAMIC_UNALIGNED = "SUPPORT_DYNAMIC_UNALIGNED"


@dataclass
class SaConfig:
    manual_unroll: bool = False
    max_unroll_times: int = 1
    only_batch_loop: bool = False
    is_nz_format: bool = False


@dataclass
class SaTileShapeConfig:
    g_tile: int
    s_kv_tile: int
    c1_tile_shape: List[int]
    v1_tile_shape: List[int]
    c2_tile_shape: List[int]
    v2_tile_shape: List[int]


def slc_attn_compute(**kwargs):
    q_nope = kwargs.get("q_nope")
    q_rope = kwargs.get("q_rope")
    k_slc = kwargs.get("k_slc")
    v_slc = kwargs.get("v_slc")
    kv_slc_act_seqs = kwargs.get("kv_slc_act_seqs")
    n_q = kwargs.get("n_q")
    n_kv = kwargs.get("n_kv")
    softmax_scale = kwargs.get("softmax_scale")
    attention_out = kwargs.get("attention_out")
    tile_config = kwargs.get("tile_config")

    dtype = q_nope.dtype
    d_n = q_nope.shape[1]
    d_r = q_rope.shape[1]
    group = n_q // n_kv

    g_tile = tile_config.g_tile
    s2_tile = tile_config.s_kv_tile
    c1_tile = tile_config.c1_tile_shape
    v1_tile = tile_config.v1_tile_shape
    c2_tile = tile_config.c2_tile_shape
    v2_tile = tile_config.v2_tile_shape

    batch_size_sym = kv_slc_act_seqs.shape[0]
    s1_n2_g_sym = q_nope.shape[0] // batch_size_sym
    s1_sym = s1_n2_g_sym // n_q
    g_loop_sym = group // g_tile

    s1_n2_s2_sym = k_slc.shape[0] // batch_size_sym
    n2_s2_sym = s1_n2_s2_sym // s1_sym
    n2_sym = n_kv

    for b_idx in pto.loop(0, batch_size_sym, 1, name="LOOP_L0_b_SA", idx_name="b_idx", 
        unroll_list=set(), submit_before_loop=True):
        for s1_idx in pto.loop(0, s1_sym, 1, name="LOOP_L1_s1_SA", idx_name="s1_idx"):
            def inside_s1_loop_sa(b_idx, s1_idx):
                cur_kv_slc_seq = pto.get_input_data(kv_slc_act_seqs, [b_idx, s1_idx])
                cur_seq = (cur_kv_slc_seq - s1_sym + 1 + s1_idx).max(0)
                cur_seq.as_intermediate_variable()
                bn_per_batch = (cur_seq + s2_tile - 1) / s2_tile

                for n2_idx in pto.loop(0, n2_sym, 1, name="LOOP_L2_n2_SA", idx_name="n2_idx"):
                    for g_idx in pto.loop(0, g_loop_sym, 1, name="LOOP_L3_g_SA", idx_name="g_idx"):
                        def inside_g_loop_sa(b_idx, s1_idx, n2_idx, g_idx):
                            cur_g_tile = g_tile
                            oi_update = pto.tensor([cur_g_tile, d_n], pto.DT_FP32, "oi_update")
                            li_update = pto.tensor([cur_g_tile, 1], pto.DT_FP32, "li_update")  
                            mi_update = pto.tensor([cur_g_tile, 1], pto.DT_FP32, "mi_update")

                            cur_offset = b_idx * s1_n2_g_sym + s1_idx * n_q + n2_idx * group + g_idx * cur_g_tile
                            oi_offset = [b_idx, s1_idx, n2_idx * group + g_idx * cur_g_tile, 0]

                            for s2_idx in pto.loop(0, bn_per_batch, 1, name="LOOP_L4_s2_SA", idx_name="s2_idx", 
                                unroll_list=pto.powers_of_2(1)):
                                def inside_s2_loop_sa(b_idx, s2_idx):
                                    cur_s2_tile = s2_tile
                                    cur_kv_offset = b_idx * s1_n2_s2_sym + s1_idx * n2_s2_sym + s2_idx * cur_s2_tile

                                    pto.set_semantic_label("Sa")
                                    qn = pto.view(q_nope, [cur_g_tile, d_n], [cur_g_tile, d_n], [cur_offset, 0])
                                    qr = pto.view(q_rope, [cur_g_tile, d_r], [cur_g_tile, d_r], [cur_offset, 0])
                                    qi = pto.tensor([cur_g_tile, d_n + d_r], dtype, "qi")
                                    pto.assemble(qn, [0, 0], qi)
                                    pto.assemble(qr, [0, d_n], qi)

                                    kj = pto.view(k_slc, [cur_s2_tile, d_n + d_r], 
                                        [(cur_seq - s2_idx * cur_s2_tile).min(cur_s2_tile), d_n + d_r],
                                        [cur_kv_offset, 0]) 
                                    vj = pto.view(v_slc, [cur_s2_tile, d_n], 
                                        [(cur_seq - s2_idx * cur_s2_tile).min(cur_s2_tile), d_n], [cur_kv_offset, 0])

                                    # C1
                                    pto.set_cube_tile_shapes([c1_tile[0], c1_tile[1]], 
                                        [c1_tile[2], c1_tile[3]], [c1_tile[4], c1_tile[5]], True)
                                    pto.set_semantic_label("Sa_QkMM")
                                    pto.set_matrix_size([qi.shape[0], 0, kj.shape[0]])
                                    sij = pto.matmul(pto.DT_FP32, qi, kj, False, True)

                                    # V1
                                    pto.set_semantic_label("Sa_Qkvec1")
                                    pto.set_vec_tile_shapes(v1_tile[0], v1_tile[1])
                                    sij_scale = pto.mul_s(sij, pto.element(sij.dtype, softmax_scale))
                                    tilda_mij = pto.row_max_single(sij_scale) 
                                    tsub = pto.sub(sij_scale, tilda_mij) 
                                    tilda_pij = pto.exp(tsub)
                                    tilda_pij_f16 = pto.cast(tilda_pij, dtype)
                                    tilda_lij = pto.row_sum_single(tilda_pij) 

                                    if pto.cond(pto.is_loop_begin(s2_idx, 0)):
                                        def inside_if_loop_begin():
                                            nonlocal oi_update, li_update, mi_update
                                            pto.set_cube_tile_shapes(
                                                [c2_tile[0], c2_tile[1]], [c2_tile[2], c2_tile[3]],
                                                [c2_tile[4], c2_tile[5]], True)
                                            pto.set_semantic_label("Sa_KvMm")
                                            pto.set_matrix_size(
                                                [tilda_pij_f16.shape[0], tilda_pij_f16.shape[1], vj.shape[1]])
                                            oi_tmp = pto.matmul(pto.DT_FP32, tilda_pij_f16, vj, False, False)
                                            pto.set_vec_tile_shapes(v2_tile[0], v2_tile[1])
                                            if pto.cond(pto.is_loop_end(s2_idx, bn_per_batch)):
                                                def inside_if_loop_end():
                                                    nonlocal oi_update
                                                    pto.set_semantic_label("Sa_KvVec2")
                                                    oi_update[:] = pto.div(oi_tmp, tilda_lij)
                                                    pto.set_vec_tile_shapes(1, 1, v2_tile[0], v2_tile[1])
                                                    oi_update_4_dim = pto.add_s(
                                                        pto.reshape(oi_update, [1, 1, cur_g_tile, d_n]), 
                                                        pto.element(oi_update.dtype, float(0)))
                                                    pto.assemble(oi_update_4_dim, oi_offset, attention_out)
                                                inside_if_loop_end()
                                            else:
                                                def inside_else_loop_end():
                                                    nonlocal oi_update
                                                    oi_update[:] = oi_tmp
                                                inside_else_loop_end()
                                            li_update[:] = tilda_lij
                                            mi_update[:] = tilda_mij
                                        inside_if_loop_begin()
                                    else:
                                        def inside_else_loop_begin():
                                            nonlocal oi_update, li_update, mi_update
                                            pto.set_semantic_label("Sa_UpdateVec2")
                                            oi = oi_update
                                            li = li_update
                                            mi = mi_update

                                            mi_new = pto.maximum(mi, tilda_mij)
                                            t1 = pto.sub(mi, mi_new)
                                            t2 = pto.exp(t1)
                                            t3 = pto.sub(tilda_mij, mi_new)
                                            t4 = pto.exp(t3)
                                            t5 = pto.mul(t4, tilda_lij)
                                            t6 = pto.mul(t2, li)
                                            li_new = pto.add(t6, t5)

                                            q3 = pto.mul(oi, t2)
                                            pto.set_cube_tile_shapes(
                                                [c2_tile[0], c2_tile[1]], [c2_tile[2], c2_tile[3]],
                                                [c2_tile[4], c2_tile[5]], True)
                                            pto.set_semantic_label("Sa_UpdateMM2")
                                            pto.set_matrix_size([tilda_pij_f16.shape[0], 
                                                tilda_pij_f16.shape[1], vj.shape[1]])
                                            q1 = pto.matmul(pto.DT_FP32, tilda_pij_f16, vj, False, False)
                                            pto.set_vec_tile_shapes(v2_tile[0], v2_tile[1])
                                            q2 = pto.mul(q1, t4)
                                            oi_tmp = pto.add(q3, q2)
                                            if pto.cond(pto.is_loop_end(s2_idx, bn_per_batch)):
                                                def inside_if_loop_end():
                                                    nonlocal oi_update
                                                    oi_update[:] = pto.div(oi_tmp, li_new)
                                                    pto.set_vec_tile_shapes(1, 1, v2_tile[0], v2_tile[1])
                                                    oi_update_4_dim = pto.add_s(
                                                        pto.reshape(oi_update, [1, 1, cur_g_tile, d_n]), 
                                                        pto.element(oi_update.dtype, float(0)))
                                                    pto.assemble(oi_update_4_dim, oi_offset, attention_out)
                                                inside_if_loop_end()
                                            else:
                                                def inside_else_loop_end():
                                                    nonlocal oi_update
                                                    oi_update[:] = oi_tmp
                                                inside_else_loop_end()
                                            li_update[:] = li_new
                                            mi_update[:] = mi_new
                                        inside_else_loop_begin()
                                inside_s2_loop_sa(b_idx, s2_idx)
                        inside_g_loop_sa(b_idx, s1_idx, n2_idx, g_idx)
            inside_s1_loop_sa(b_idx, s1_idx)


def slc_attn(**kwargs):
    q_nope = kwargs.get("q_nope")
    q_rope = kwargs.get("q_rope")
    k_slc = kwargs.get("k_slc")
    v_slc = kwargs.get("v_slc")
    kv_slc_act_seqs = kwargs.get("kv_slc_act_seqs")
    n_q = kwargs.get("n_q")
    n_kv = kwargs.get("n_kv")
    softmax_scale = kwargs.get("softmax_scale")
    attention_out = kwargs.get("attention_out")
    tile_config = kwargs.get("tile_config")

    with pto.function("SA_MAIN", [q_nope, q_rope, k_slc, v_slc, kv_slc_act_seqs], [attention_out]):
        slc_attn_compute(
            q_nope=q_nope,
            q_rope=q_rope,
            k_slc=k_slc,
            v_slc=v_slc,
            kv_slc_act_seqs=kv_slc_act_seqs,
            n_q=n_q,
            n_kv=n_kv,
            softmax_scale=softmax_scale,
            attention_out=attention_out,
            tile_config=tile_config
        )


def test_sa_ut(input_param, tile_config, config):
    d_type = pto.DT_FP16

    b = input_param[0]
    sq = input_param[1]
    nq = input_param[2]
    nkv = input_param[3]
    dn = input_param[4]
    dr = input_param[5]
    smax = input_param[6]
    softmax_scale = float(1.0 / math.sqrt((dn + dr)))

    kv_format = pto.TILEOP_NZ if config.is_nz_format else pto.TILEOP_ND

    q_nope_shape = [b * sq * nq, dn]
    q_rope_shape = [b * sq * nq, dr]
    k_slc_shape = [b * sq * nkv * smax, dn + dr]
    v_slc_shape = [b * sq * nkv * smax, dn]
    sa_out_shape = [b, sq, nq, dn]

    act_seqs = pto.tensor([b, sq], pto.DT_INT32, "act_seqs")
    q_nope = pto.tensor(q_nope_shape, d_type, "q_nope")
    q_rope = pto.tensor(q_rope_shape, d_type, "q_rope")
    k_slc = pto.tensor(k_slc_shape, d_type, "k_slc", kv_format)
    v_slc = pto.tensor(v_slc_shape, d_type, "v_slc", kv_format)
    sa_out = pto.tensor(sa_out_shape, pto.DT_FP32, "sa_out")

    slc_attn(
        q_nope=q_nope,
        q_rope=q_rope,
        k_slc=k_slc,
        v_slc=v_slc,
        kv_slc_act_seqs=act_seqs,
        n_q=nq,
        n_kv=nkv,
        softmax_scale=softmax_scale,
        attention_out=sa_out,
        tile_config=tile_config
    )


def main():
    pto.set_codegen_config(KEY_SUPPORT_DYNAMIC_UNALIGNED, True)

    tile_config = SaTileShapeConfig(
        g_tile=128,
        s_kv_tile=1024,
        c1_tile_shape=[128, 1024, 64, 64, 128, 128],
        v1_tile_shape=[16, 256],
        c2_tile_shape=[128, 1024, 64, 64, 128, 128],
        v2_tile_shape=[16, 256]
    )

    input_param = [32, 1, 128, 1, 512, 64, 1024]
    config = SaConfig()
    test_sa_ut(input_param, tile_config, config)

if __name__ == "__main__":
    logging.basicConfig(format='%(asctime)s - %(filename)s:%(lineno)d - %(levelname)s: %(message)s', level=logging.INFO)
    main()
